File: services/api/src/models/checkout.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::{FromRow, Type};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default, Type)]
#[sqlx(type_name = "checkout_status", rename_all = "UPPERCASE")]
#[serde(rename_all = "UPPERCASE")]
pub enum CheckoutStatus {
    #[default]
    Active,
    Returned,
    Overdue,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Checkout {
    pub id: Uuid,
    pub user_id: Uuid,
    pub book_id: Uuid,
    pub status: CheckoutStatus,
    pub checked_out_at: DateTime<Utc>,
    pub due_date: DateTime<Utc>,
    pub returned_at: Option<DateTime<Utc>>,
    pub renewal_count: i32,
    pub max_renewals: i32,
    pub overdue_email_sent: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CheckoutWithDetails {
    #[serde(flatten)]
    pub checkout: Checkout,
    pub user: CheckoutUser,
    pub book: CheckoutBook,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CheckoutUser {
    pub id: Uuid,
    pub name: String,
    pub email: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CheckoutBook {
    pub id: Uuid,
    pub title: String,
    pub author: String,
    pub isbn: String,
}

#[derive(Debug, Deserialize)]
pub struct CreateCheckoutRequest {
    pub user_id: Uuid,
    pub book_id: Uuid,
    pub due_date: Option<DateTime<Utc>>,
}

#[derive(Debug, Deserialize)]
pub struct CheckoutBookRequest {
    pub isbn: String,
    pub user_id: Uuid,
}

#[derive(Debug, Deserialize)]
pub struct ReturnBookRequest {
    pub checkout_id: Uuid,
}

#[derive(Debug, Deserialize)]
pub struct RenewCheckoutRequest {
    pub checkout_id: Uuid,
}

#[derive(Debug, Deserialize)]
pub struct CheckoutSearchQuery {
    pub user_id: Option<Uuid>,
    pub book_id: Option<Uuid>,
    pub status: Option<CheckoutStatus>,
    pub overdue: Option<bool>,
    pub limit: Option<i64>,
    pub offset: Option<i64>,
}

File: packages/shared/src/types/checkout.ts
import { z } from "zod";

export const CheckoutStatusSchema = z.enum(["ACTIVE", "RETURNED", "OVERDUE"]);

export const CheckoutSchema = z.object({
  id: z.string().uuid(),
  userId: z.string().uuid(),
  bookId: z.string().uuid(),
  status: CheckoutStatusSchema.default("ACTIVE"),
  checkedOutAt: z.date(),
  dueDate: z.date(),
  returnedAt: z.date().optional(),
  renewalCount: z.number().int().min(0).default(0),
  maxRenewals: z.number().int().min(0).default(2),
  overdueEmailSent: z.boolean().default(false),
  createdAt: z.date(),
  updatedAt: z.date(),
});

export const CreateCheckoutSchema = z.object({
  userId: z.string().uuid(),
  bookId: z.string().uuid(),
  dueDate: z.date().optional(),
});

export const CheckoutBookSchema = z.object({
  isbn: z.string().min(10).max(13),
});

export const ReturnBookSchema = z.object({
  checkoutId: z.string().uuid(),
});

export const RenewCheckoutSchema = z.object({
  checkoutId: z.string().uuid(),
});

export const CheckoutSearchQuerySchema = z.object({
  userId: z.string().uuid().optional(),
  bookId: z.string().uuid().optional(),
  status: CheckoutStatusSchema.optional(),
  overdue: z.boolean().optional(),
  limit: z.number().int().min(1).max(100).default(20),
  offset: z.number().int().min(0).default(0),
});

export const CheckoutWithDetailsSchema = CheckoutSchema.extend({
  user: z.object({
    id: z.string().uuid(),
    name: z.string(),
    email: z.string().email(),
  }),
  book: z.object({
    id: z.string().uuid(),
    title: z.string(),
    author: z.string(),
    isbn: z.string(),
  }),
});

export type CheckoutStatus = z.infer<typeof CheckoutStatusSchema>;
export type Checkout = z.infer<typeof CheckoutSchema>;
export type CreateCheckout = z.infer<typeof CreateCheckoutSchema>;
export type CheckoutBook = z.infer<typeof CheckoutBookSchema>;
export type ReturnBook = z.infer<typeof ReturnBookSchema>;
export type RenewCheckout = z.infer<typeof RenewCheckoutSchema>;
export type CheckoutSearchQuery = z.infer<typeof CheckoutSearchQuerySchema>;
export type CheckoutWithDetails = z.infer<typeof CheckoutWithDetailsSchema>;
